core/engine/src/object/builtins/lazy_builtin.rs

```diff
@@ -0,0 +1,451 @@
+use super::JsFunction;
+use crate::{
+    gc::custom_trace,
+    object::{
+        internal_methods::{
+            non_existant_call, non_existant_construct, ordinary_define_own_property,
+            ordinary_delete, ordinary_get, ordinary_get_own_property, ordinary_get_prototype_of,
+            ordinary_has_property, ordinary_is_extensible, ordinary_own_property_keys,
+            ordinary_prevent_extensions, ordinary_set, ordinary_set_prototype_of, ordinary_try_get,
+            CallValue, InternalMethodContext, InternalObjectMethods,
+        },
+        JsPrototype,
+    },
+    property::{PropertyDescriptor, PropertyKey},
+    realm::{Realm, RealmInner},
+    Context, JsData, JsNativeError, JsObject, JsResult, JsValue,
+};
+use boa_gc::{Finalize, Trace, WeakGc};
+use boa_string::JsString;
+use std::cell::Cell;
+
+#[derive(Debug, Clone, Trace, Finalize)]
+pub(crate) enum BuiltinKind {
+    Function(JsFunction),
+    Ordinary,
+}
+
+/// A builtin function. Used for lazy initialization of builtins.
+
+#[derive(Clone, Finalize)]
+pub struct LazyBuiltIn {
+    pub(crate) init: fn(&Realm),
+    pub(crate) is_initialized: Cell<bool>,
+    pub(crate) kind: BuiltinKind,
+    pub(crate) realm_inner: Option<WeakGc<RealmInner>>,
+    pub(crate) name: JsString,
```

Comment:
We can put `init` and `realm_inner` into a single `Option` and remove `is_initialized`. That way, we can use `Option::take` (which will set the option to `None` on access) to determine if the builtin has already been initialized; if `Option::take` returns `Some`, initialize. Else, don't initialize.

Reply:
Done